compiler/rustc_passes/src/dead.rs

```diff
@@ -677,30 +677,91 @@ impl<'tcx> DeadVisitor<'tcx> {
         let inherent_impls = self.tcx.inherent_impls(def_id);
         for &impl_did in inherent_impls.iter() {
             for item_did in self.tcx.associated_item_def_ids(impl_did) {
-                if let Some(def_id) = item_did.as_local() {
-                    if self.live_symbols.contains(&def_id) {
-                        return true;
-                    }
+                if let Some(def_id) = item_did.as_local()
+                    && self.live_symbols.contains(&def_id)
+                {
+                    return true;
                 }
             }
         }
         false
     }
 
-    fn warn_dead_code(
+    fn store_dead_field_or_variant(
         &mut self,
         id: hir::HirId,
         span: rustc_span::Span,
         name: Symbol,
+    ) {
+        let field_or_variant_local_def_id = |parent_node, node| -> Option<(LocalDefId, bool)> {
+            match (parent_node, node) {
+                (
+                    Node::Item(hir::Item {
+                        kind:
+                            hir::ItemKind::Struct(hir::VariantData::Struct(..), ..)
+                            | hir::ItemKind::Union(hir::VariantData::Struct(..), ..),
+                        def_id,
+                        ..
+                    }),
+                    Node::Field(..),
+                ) => Some((*def_id, true)),
+                (
+                    Node::Item(hir::Item { kind: hir::ItemKind::Enum(..), def_id, .. }),
+                    Node::Variant(..),
+                ) => Some((*def_id, false)),
+                (
+                    Node::Variant(hir::Variant { data: hir::VariantData::Struct(..), id, .. }, ..),
+                    Node::Field(..),
+                ) => self
+                    .tcx
+                    .hir()
+                    .find_parent_node(*id)
+                    .and_then(|id| self.tcx.hir().opt_local_def_id(id).map(|id| (id, true))),
+                _ => None,
+            }
+        };
+        if let Some(parent_hir_id) = self.tcx.hir().find_parent_node(id)
+            && let Some(parent_node) = self.tcx.hir().find(parent_hir_id)
+            && let Some(node) = self.tcx.hir().find(id)
+            && let Some((did, is_field)) = field_or_variant_local_def_id(parent_node, node)
+            && !name.as_str().starts_with('_')
+        {
+            if is_field {
+                self.dead_fields.entry(did).or_default().push((id, span, name));
```

Comment:
These are accumulated in a member that is (I think?) shared for all of the items in a module, but what happens when we have two structs both with dead fields? Does that mess with the diagnostics?

```rust
struct A {
  a: i32,
}

struct B {
  b: i32,
}

Reply:
In this case, warning looks like the following.

```
error: field is never read: `a`
 --> src/main.rs:5:5
  |
5 |     a: i32,
  |     ^^^^^^
  |
note: the lint level is defined here
 --> src/main.rs:1:11
  |
1 | #![forbid(dead_code)]
  |           ^^^^^^^^^
  = note: `A` has a derived impl for the trait `Debug`, but this is intentionally ignored during dead code analysis

error: field is never read: `b`
  --> src/main.rs:10:5
   |
10 |     b: i32,
   |     ^^^^^^
   |
   = note: `B` has a derived impl for the trait `Debug`, but this is intentionally ignored during dead code analysis

error: could not compile `debug_playground` due to 2 previous errors
```

Reply:
oh, didn't notice it was a `Map<DefId, Vec<HirId>>`